scripts/rdp-ntlm-info.nse

```diff
@@ -0,0 +1,168 @@
+-- <elem key="DNS_Computer_Name">W16GA-SRV01.W2016.lab</elem>
+-- <elem key="DNS_Tree_Name">W2016.lab</elem>
+-- <elem key="Product_Version">10.0.14393</elem>
+-- <elem key="System_Time">2019-06-13T10:38:35+00:00</elem>
+
+
+
+author = "Tom Sellers"
+license = "Same as Nmap--See https://nmap.org/book/man-legal.html"
+categories = {"default", "discovery", "safe"}
+
+portrule = shortport.port_or_service(3389, "ms-wbt-server")
+
+action = function(host, port)
+
+  local comm = rdp.Comm:new(host, port)
+  if ( not(comm:connect()) ) then
+    return false, fail("Failed to connect to server")
+  end
+
+  -- Request CredSSP protocol = 3
+  local cr = rdp.Request.ConnectionRequest:new(11)
+  local status, _ = comm:exch(cr)
+  if ( not(status) ) then
+    comm:close()
+    return
+  end
+
+  -- This script could include code to detect which security protocols the
+  -- target claims that it accepts however that's less than useful because
+  -- 1. Windows XP doesn't provide that layer of the packet
+  -- 2. Even when configured for RDP Security only, Windows Server 2008
+  --    will still let you connect over TLS and start the CredSSP nego.
+
+  local _, response, recvtime
+  status, _ = comm.socket:reconnect_ssl()
+  if status then
+    stdnse.debug1("Sending NTLM NEGOTIATE..")
+
+    -- NTLMSSP Negotiate request mimicking a Windows 10 client
+    local NTLM_NEGOTIATE_BLOB = stdnse.fromhex(
+      "30 37 A0 03 02 01 60 A1 30 30 2E 30 2C A0 2A 04 28" ..
+      "4e 54 4c 4d 53 53 50 00" .. -- Identifier - NTLMSSP
+      "01 00 00 00" ..  -- Type: NTLMSSP Negotiate - 01
+      "B7 82 08 E2 " .. -- Flags (NEGOTIATE_SIGN_ALWAYS | NEGOTIATE_NTLM | NEGOTIATE_SIGN | REQUEST_TARGET | NEGOTIATE_UNICODE)
+      "00 00 " ..       -- DomainNameLen
+      "00 00" ..        -- DomainNameMaxLen
+      "00 00 00 00" ..  -- DomainNameBufferOffset
+      "00 00 " ..       -- WorkstationLen
+      "00 00" ..        -- WorkstationMaxLen
+      "00 00 00 00" ..  -- WorkstationBufferOffset
+      "0A" ..           -- ProductMajorVersion = 10
+      "00 " ..          -- ProductMinorVersion = 0
+      "63 45 " ..       -- ProductBuild = 0x4563 = 17763
+      "00 00 00" ..     -- Reserved
+      "0F"              -- NTLMRevision = 5 = NTLMSSP_REVISION_W2K3
+    )
+
+    -- Not using comm:exch here since that performs some processing on the
+    -- packet that isn't appropriate in this case.
+    status, response = comm:send(NTLM_NEGOTIATE_BLOB)
+    if ( not(status) ) then
+      return false, response
+    end
+
+    status, response = comm:recv()
+    if status then
+      recvtime = os.time()
+    end
+  else
+    comm:close()
+    stdnse.debug1("Unable to establish a TLS connection which is required to negotiation CredSSP.")
+    return
+  end
+
+  if response == nil then
+    return
+  end
+
+  -- Continue only if NTLMSSP response is returned
+  local start = response:find("NTLMSSP")
+  response = response:sub(start)
+  if not string.match(response, "^NTLMSSP") then
```

Comment:
If "NTLMSSP" is not in `response`, then `start` will be `nil`, which we can check instead of using `string.match()` here. If it *is* in `response`, then after the `sub()` call above, this `string.match()` will always be true.